File: lab-1/lab1_1.js
class Animal {
  makeSound() {
    throw new Error(
      "Метод makeSound() повинен бути реалізований у похідному класі."
    );
  }

  sleep() {
    console.log(`${this.constructor.name} is sleeping...`);
  }
}

class Dog extends Animal {
  makeSound() {
    console.log("Woof! Woof!");
  }
}

class Cat extends Animal {
  makeSound() {
    console.log("Meow! Meow!");
  }
}

const dog = new Dog();
const cat = new Cat();

dog.makeSound();
cat.makeSound();

dog.sleep();
cat.sleep();

File: lab-1/lab1-1.py
from abc import ABC, abstractmethod

class Animal(ABC):
  @abstractmethod
  def make_sound(self):
    pass

  def sleep(self):
    print(f"{self.__class__.__name__} is sleeping...")

class Dog(Animal):
  def make_sound(self):
    print("Woof! Woof!")

class Cat(Animal):
  def make_sound(self):
    print("Meow! Meow!")

if __name__ == "__main__":
  dog = Dog()
  cat = Cat()

  dog.make_sound()
  cat.make_sound()
  
  dog.sleep()
  cat.sleep()
